src/compiler/parser.ts

```diff
@@ -2615,6 +2615,7 @@ namespace ts {
         }
 
         function parseUnionOrIntersectionType(kind: SyntaxKind, parseConstituentType: () => TypeNode, operator: SyntaxKind): TypeNode {
+            parseOptional(operator);

```

Comment:
Might be a good idea. For example, looking at where `isStartOfType` is called, we wouldn't properly handle a tuple type `[| 0 | 1, | "foo" | "bar"]` without this change, so maybe add a test for that.